File: java/src/main/java/structural/Adapter.java
package structural;

// Expected interface
interface Logger {

    void info(String msg);
}

// Legacy library
class OldLogger {

    public void write(String text) {
        System.out.println("[OLD LOG] " + text);
    }
}

// Adapter — maps the old API onto the required interface
class LoggerAdapter implements Logger {

    private final OldLogger old;

    public LoggerAdapter(OldLogger old) {
        this.old = old;
    }

    @Override
    public void info(String msg) {
        old.write("INFO: " + msg); // convert to legacy format
    }
}

@SuppressWarnings("unused")
final class AdapterExample {

    private AdapterExample() {
    }

    public static void example() {
        Logger logger = new LoggerAdapter(new OldLogger());
        logger.info("Started"); // Client never touches OldLogger directly
    }
}

File: rust/src/structural/adapter.rs
struct OldLogger;
impl OldLogger {
    fn log(&self, msg: &str) {
        println!("[OLD] {}", msg);
    }
}

// New interface required by the system:
trait Logger {
    fn write(&self, message: &str);
}

// Adapter that wraps OldLogger with the new interface:
struct LoggerAdapter {
    old_logger: OldLogger,
}
impl Logger for LoggerAdapter {
    fn write(&self, message: &str) {
        // Forward the call to the legacy method:
        self.old_logger.log(message);
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn example() {
        // Using the adapter:
        let adapter = LoggerAdapter {
            old_logger: OldLogger,
        };
        adapter.write("Some important information");
    }
}
